File: android/src/main/java/com/terrylinla/rnsketchcanvas/utils/layers/Layer.java
package com.terrylinla.rnsketchcanvas.utils.layers;

import android.support.annotation.FloatRange;

public class Layer {

    /**
     * rotation relative to the layer center, in degrees
     */
    @FloatRange(from = 0.0F, to = 360.0F)
    private float mRotationInDegrees;

    private float mScale;
    /**
     * top left X coordinate, relative to parent canvas
     */
    private float x;
    /**
     * top left Y coordinate, relative to parent canvas
     */
    private float y;
    /**
     * is layer flipped horizontally (by X-coordinate)
     */
    private boolean mIsFlipped;

    public Layer() {
        reset();
    }

    protected void reset() {
        this.mRotationInDegrees = 0.0F;
        this.mScale = 1.0F;
        this.mIsFlipped = false;
        this.x = 0.0F;
        this.y = 0.0F;
    }

    public void postScale(float scaleDiff) {
        float newVal = mScale + scaleDiff;
        if (newVal >= getMinScale() && newVal <= getMaxScale()) {
            mScale = newVal;
        }
    }

    protected float getMaxScale() {
        return Limits.MAX_SCALE;
    }

    protected float getMinScale() {
        return Limits.MIN_SCALE;
    }

    public void postRotate(float rotationInDegreesDiff) {
        this.mRotationInDegrees += rotationInDegreesDiff;
        this.mRotationInDegrees %= 360.0F;
    }

    public void postTranslate(float dx, float dy) {
        this.x += dx;
        this.y += dy;
    }

    public void flip() {
        this.mIsFlipped = !mIsFlipped;
    }

    public float initialScale() {
        return Limits.INITIAL_ENTITY_SCALE;
    }

    public float getRotationInDegrees() {
        return mRotationInDegrees;
    }

    public void setRotationInDegrees(@FloatRange(from = 0.0, to = 360.0) float rotationInDegrees) {
        this.mRotationInDegrees = rotationInDegrees;
    }

    public float getScale() {
        return mScale;
    }

    public void setScale(float scale) {
        this.mScale = scale;
    }

    public float getX() {
        return x;
    }

    public void setX(float x) {
        this.x = x;
    }

    public float getY() {
        return y;
    }

    public void setY(float y) {
        this.y = y;
    }

    public boolean isFlipped() {
        return mIsFlipped;
    }

    public void setFlipped(boolean flipped) {
        mIsFlipped = flipped;
    }

    interface Limits {
        float MIN_SCALE = 0.06F;
        float MAX_SCALE = 4.0F;
        float INITIAL_ENTITY_SCALE = 0.4F;
    }
}

File: ios/RNSketchCanvas/RNSketchCanvas/entities/layer/Layer.m
//
//  Layer.m
//  RNSketchCanvas
//
//  Created by Thomas Steinbrüchel on 23.10.18.
//  Copyright © 2018 Terry. All rights reserved.
//

#import "Layer.h"

@implementation Layer

{
    CGFloat rotationInDegrees;
    CGFloat scale;
    CGFloat x;
    CGFloat y;
    BOOL isFlipped;
    Limits *limits;
}

- (instancetype)initAndReset {
    self = [super init];
    if (self) {
        rotationInDegrees = 0.0;
        scale = 1.0;
        isFlipped = FALSE;
        x = 0.0;
        y = 0.0;
        limits = [[Limits init] initAndSetup];
    }
    return self;
}

- (CGFloat)getMinScale {
    return limits.MIN_SCALE;
}

- (CGFloat)getMaxScale {
    return limits.MAX_SCALE;
}

- (CGFloat)getInitialScale {
    return limits.INITIAL_ENTITY_SCALE;
}

- (void)postScale:(CGFloat)scaleDiff {
    CGFloat newVal = scale + scaleDiff;
    if (newVal >= [self getMinScale] && newVal <= [self getMaxScale]) {
        scale = newVal;
    }
}

- (void)postRotate:(CGFloat)rotationInDegreesDiff {
    rotationInDegrees += rotationInDegreesDiff;
    rotationInDegrees = fmod(rotationInDegrees, 360.0);
}

- (void)postTranslate:(CGFloat)dx dy:(CGFloat)dy {
    x += dx;
    y += dy;
}

- (void)flip {
    isFlipped = !isFlipped;
}

- (CGFloat)getRotationInDegrees {
    return rotationInDegrees;
}

- (void)setRotationInDegrees:(CGFloat)rotationInDegrees {
    self.rotationInDegrees = rotationInDegrees;
}

- (CGFloat)getScale {
    return scale;
}

- (void)setScale:(CGFloat)scale {
    self.scale = scale;
}

- (CGFloat)getX {
    return x;
}

- (void)setX:(CGFloat)x {
    self.x = x;
}

- (CGFloat)getY {
    return y;
}

- (void)setY:(CGFloat)y {
    self.y = y;
}

- (BOOL)isFlipped {
    return isFlipped;
}

- (void)setFlipped:(BOOL)flipped {
    isFlipped = flipped;
}
@end



@implementation Limits
- (instancetype)initAndSetup {
    self = [super init];
    if (self) {
        _MIN_SCALE = 0.06;
        _MAX_SCALE = 4.0;
        _INITIAL_ENTITY_SCALE = 0.4;
    }
    return self;
}
@end
